### DISCUSSION TITLE
Importing ethr DID and set controllerKeyId for sending a transaction.

### DISCUSSION THREAD

**Discussion by chike0905** [decentralized-identity/veramo#779] [Q&A]
I try to import EthrDID that is controlled by a pre-generated private key and add another key to the DID Document.  I tried like following;
```typescript
const agent = setup();

const key: MinimalImportableKey = {
  kms: 'local',
  privateKeyHex: MY_PRIVATE_KEY,
  type: 'Secp256k1',
}
const veramoDid: MinimalImportableIdentifier = {
  keys: [key],
  provider: 'did:ethr',
  did: "did:ether:1337:" + ADDRESS
};

res = await agent.didManagerImport(veramoDid)
  .catch((e: any) => {
    console.error(e);
    return;
  }) as IIdentifier;

const newKey = await agent.keyManagerCreate({ kms: 'local', type: 'Secp256k1' })
  .catch((e: any) => {
    console.error(e);
    return;
  }) as IKey;

await agent.didManagerAddKey({ did: veramoDid.did, key: newKey, options: {} })
  .catch((e: any) => {
    console.error(e);
    return;
  });
```

I got an error like this.
```
identifier does not list a `controllerKeyId`
```

it seems that the imported DID has not `controllerKeyId` and ethr-did-provider uses it for signing a transaction to update the did document. Hence, I got the same error in `agent.didManagerAddService`. How can I set the value in the importing workflow?

**Comment by chike0905** [comment]
Sorry, I solved it by myself.
We can set `controllerKeyId ` by setting it in the object for didManagerImport (`MinimalImportableIdentifier`).
Like this,
```
const veramoDid: MinimalImportableIdentifier = {
  keys: [key],
  provider: 'did:ethr',
  did: "did:ethr:1337:" + ADDRESS,
  controllerKeyId: KEY_ID
};
```

But, to do this, you should get `KEY_ID` for the private key before importing the DID.
So, as a user workflow, I think it is better that didManagerImport assigns the parameter automatically in `did-manager/src/id-manager.ts`.

**Reply by mirceanis** [reply]
Great to hear that you solved it!

Indeed, the `didManagerImport` can be improved with the functionality you suggested, and more..
It probably also needs to be able to bootstrap from an already imported existing key, for DID methods that wrap a key like did:ethr and did:key.

We encourage contributions from our community, so if you think you can make any of these improvements, feel free to propose a PR, or even phrase it as an issue.

**Reply by chike0905** [reply]
@mirceanis  Thank you for replaying.
As you have mentioned, it should be considered importing a key (or importing a did that is tied with a keypair). I think it is related to an interoperability issue of DID methods. 

As an additional question, how does the key-manager with a default setting take a key id?   In my opinion, how to manage and identify keys in a veramo agent is essential to keep interoperability of each DID method.

**Reply by mirceanis** [reply]
Actually, the key IDs (`IKey.kid`) in Veramo do not match the IDs of their corresponding public keys in DID documents.
This is mostly because one key can appear in multiple DID documents and each DID method can use its own scheme for generating key IDs.

However, if you already know the DID document key id, you can use it in your import, if that makes it easier for you.

```ts
const key: MinimalImportableKey = {
  kms: 'local',
  privateKeyHex: MY_PRIVATE_KEY,
  type: 'Secp256k1',
  kid: ID_FROM_DID_DOCUMENT
}
const veramoDid: MinimalImportableIdentifier = {
  keys: [key],
  // ...
}
```

Also, in the next version of Veramo we are preparing [some utility methods](https://github.com/uport-project/veramo/blob/5202ef1dd51fd3b7ad57591fd726fdc571bc8492/packages/utils/src/did-utils.ts#L66) to help you match the managed keys of a Veramo identifier with the corresponding entries from its DID document:

```ts
import { mapIdentifierKeysToDoc } from '@veramo/utils'

//...
const extendedKeys = mapIdentifierKeysToDoc(veramoDID, 'authentication', { agent })
console.log(extendedKeys[0].meta.verificationMethod.id)
```

This automatically resolves the DID document of the identifier and matches the public key material from the DID document to the public key material from the `keyManager`.
This is very early, so your feedback regarding this API is very welcome ;)
